_posts/2018-26-04-elixir-post.md

```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title:  "Working With Object Oriented and Functional Programming Without Losing Your Mind"
+date:   2018-04-26 12:00:00 -0400
+categories: elixir coding ruby oo functional
+author: Tania Paiva
+---
+
+## Working With Object Oriented and Functional Programming Without Losing Your Mind
+
+At Forward Financing we work with several technologies, our stack is pretty mixed.
+We work with Rails, Phoenix, React, Redux among others, which means we are changing the
+development environment sometimes more than once at day.
+
+Sometimes in the morning you are developing a new feature in the Elixir-Phoenix environment, and in the
+afternoon we are improving existing code made on Ruby on Rails.
+
+The main thing with this "switching" is the programming paradigm that each of this technologies uses is totally
+different.
+
+### Object Oriented Programming (OOP) and Functional Programming (FP)
+
+A program has two main components, the data and the behavior.
+The main difference between OOP and FP is how we manage these components.
+
+OOP puts together data and its associated behavior in a single place called an `Object`. For FP, the data and behavior are  different things and should be separated.
+
+FP tries to avoid sharing state and mutation on data. The objective of FP is building pure functions with no side effects.
+
+OOP provides properties and methods for objects, they live inside the class structure, an object has a class and we can
+have an instance of it.
+
+A function is a piece of code that is called by name. It can be passed data to operate on (parameters) and can optionally return data (the most common practice is to return something). All data that is passed to a function is explicitly passed.
+
+A method is a piece of code that is called by a name that is associated with the Object.
+
+Let's see a practical example, we need to implement a program that shuffles cards on a deck.
+
+### Object Oriented approach
+``` ruby
+class Card
+  def new(value)
+    @value = value
+  end
+
+  def value
+    return @value
+  end
+
+end
+
+class Deck
+  def new(values)
+    @cards = []
+    values.each |v| do
+      @cards << Card.new(v)
+    end
+  end
+
+  def shuffle_deck
+    @cards = @cards.shuffle
+  end
+
+  def contains?(card_value)
```

Comment:
I would probably get rid of the explicit `true` statement
```  
def contains?(card_value)
    @cards.each do |c|
        return c.value == card_value
      end
    false
  end
```
Or to bring it closer to the elixir example:

 ```
def contains?(card_value)
    @cards.any? do |card|
      card.value == card_value
    end
  end
```